crates/planner/src/logical/expr.rs

```diff
@@ -0,0 +1,317 @@
+        Self::Select(Box::new(Select { input, vars, exprs }))
+    }
+
+    /// The type of a relation expression
+    pub fn ty(&self) -> &Type {
+        match self {
+            Self::RelVar(_, ty) => ty,
+            Self::Select(op) => op.input.ty(),
+            Self::Proj(op) => &op.ty,
+            Self::Join(_, ty) => ty,
+            Self::Union(a, _) | Self::Minus(a, _) | Self::Dedup(a) => a.ty(),
+        }
+    }
+}
+
+/// A list of bound variables and their types.
+/// Used as a typing context scoped to an expression.
+#[derive(Debug, Clone, Default)]
+pub struct Vars(Vec<(String, Type)>);
+
+impl From<&TableSchema> for Vars {
+    fn from(value: &TableSchema) -> Self {
+        Self(
+            value
+                .columns()
+                .iter()
+                .map(|schema| (schema.col_name.to_string(), Type::Alg(schema.col_type.clone())))
+                .collect(),
+        )
+    }
+}
+
+impl From<Vec<(String, Type)>> for Vars {
+    fn from(vars: Vec<(String, Type)>) -> Self {
+        Self(vars)
+    }
+}
+
+impl<I: Into<(String, Type)>> FromIterator<I> for Vars {
+    fn from_iter<T: IntoIterator<Item = I>>(iter: T) -> Self {
+        Self(iter.into_iter().map(|item| item.into()).collect())
+    }
+}
+
+impl Vars {
+    /// Add a new variable binding to the list
+    pub fn push(&mut self, var: (String, Type)) {
+        self.0.push(var)
+    }
+
+    /// Returns an iterator over the variable names and types
+    pub fn iter(&self) -> impl Iterator<Item = (&str, &Type)> {
+        self.0.iter().map(|(name, ty)| (name.as_str(), ty))
+    }
+
+    /// Find a variable by name in this list.
+    /// Returns its type and position in the list if found.
+    pub fn find(&self, param: &str) -> Option<(usize, &Type)> {
+        self.0
+            .iter()
+            .enumerate()
+            .find(|(_, (name, _))| name == param)
+            .map(|(i, (_, ty))| (i, ty))
+    }
+
+    /// Find a variable from the list of in scope variables.
+    /// Return its type and position in the list if found.
+    /// Return an error otherwise.
+    pub fn expect_var(&self, param: &str, expected: Option<&Type>) -> TypingResult<(usize, &Type)> {
+        self.find(param)
+            // Return resolution error if param not in scope
+            .ok_or_else(|| ResolutionError::unresolved_var(param).into())
+            // Return type error if param in scope but wrong type
+            .and_then(|(i, ty)| match expected {
+                Some(expected) if ty != expected => Err(ConstraintViolation::eq(expected, ty).into()),
+                _ => Ok((i, ty)),
+            })
+    }
+
+    /// Find an in scope table variable and field.
+    /// Return its type, variable and field positions if found.
+    /// Return an error otherwise.
+    pub fn expect_field(
+        &self,
+        table: &str,
+        field: &str,
+        expected: Option<&Type>,
+    ) -> TypingResult<(usize, usize, &AlgebraicType)> {
+        self.find(table)
```

Comment:
I think is better to make a struct that describes what are the `usize, usize` here...